pkg/gatewayserver/io/ws/ws.go

```diff
@@ -269,6 +280,16 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 					conn.Disconnect(err)
 					return
 				}
+				if gatewayCanSendPong {
+					missedPongsMu.Lock()
+					missedPongs++
+					if missedPongs == s.cfg.MissedPongThreshold {
+						logger.WithError(errMissedTooManyPongs).Warn("Disconnect gateway")
+						conn.Disconnect(errMissedTooManyPongs)
+						return
+					}
+					missedPongsMu.Unlock()
```

Comment:
`Unlock` after increment, otherwise you don't unlock when you disconnect the gateway. Also `.New`.

Reply:
The mutex is local to the function stack and will get GCed when we exit the function but this is good.

Reply:
If due to some cosmic misfortune `ReadMessage` was running the pong handler while you return an error here, the main goroutine would not have returned, wouldn't it ?

Reply:
That's exceedingly rare but not impossible, so it's good to do this.